File: rate_limit_guard.swift
#!/usr/bin/env swift

import Foundation

let threshold = 90
let verbose = CommandLine.arguments.contains("--verbose") || CommandLine.arguments.contains("-v")
let noSleep = CommandLine.arguments.contains("--no-sleep")

// Check for bypass
if ProcessInfo.processInfo.environment["CLAUDE_NO_LIMIT"] == "1" {
    if verbose { print("⚠️  Rate limit guard bypassed (CLAUDE_NO_LIMIT=1)") }
    exit(0)
}

// Read session key from environment
guard let sessionKey = ProcessInfo.processInfo.environment["CLAUDE_SESSION_KEY"] else {
    if verbose { print("⚠️  No CLAUDE_SESSION_KEY set - skipping rate limit check") }
    exit(0)
}

// Read org ID from environment
guard let orgId = ProcessInfo.processInfo.environment["CLAUDE_ORG_ID"] else {
    if verbose { print("⚠️  No CLAUDE_ORG_ID set - skipping rate limit check") }
    exit(0)
}

func fetchUsageData(sessionKey: String, orgId: String) async throws -> (utilization: Int, resetsAt: Date?) {
    guard let url = URL(string: "https://claude.ai/api/organizations/\(orgId)/usage") else {
        throw NSError(domain: "ClaudeAPI", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
    }

    var request = URLRequest(url: url)
    request.setValue("sessionKey=\(sessionKey)", forHTTPHeaderField: "Cookie")
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.httpMethod = "GET"
    request.timeoutInterval = 10

    let (data, response) = try await URLSession.shared.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse,
          httpResponse.statusCode == 200 else {
        throw NSError(domain: "ClaudeAPI", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to fetch"])
    }

    if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
       let fiveHour = json["five_hour"] as? [String: Any],
       let utilization = fiveHour["utilization"] as? Int {

        var resetDate: Date? = nil
        if let resetsAt = fiveHour["resets_at"] as? String {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            resetDate = formatter.date(from: resetsAt)
        }

        return (utilization, resetDate)
    }

    throw NSError(domain: "ClaudeAPI", code: 4, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
}

Task {
    do {
        let (utilization, resetsAt) = try await fetchUsageData(sessionKey: sessionKey, orgId: orgId)

        if verbose {
            print("✓ Usage: \(utilization)% (threshold: \(threshold)%)")
        }

        if utilization >= threshold && !noSleep {
            // Calculate sleep time until reset
            var sleepSeconds = 600 // Default 10 minutes

            if let resetDate = resetsAt {
                let secondsUntilReset = Int(resetDate.timeIntervalSinceNow) + 60 // Add 1 min buffer
                if secondsUntilReset > 0 {
                    sleepSeconds = secondsUntilReset
                }
            }

            let minutes = sleepSeconds / 60
            print("⚠️  Claude usage at \(utilization)% - sleeping \(minutes) minutes until reset...")

            // Actually sleep
            Thread.sleep(forTimeInterval: Double(sleepSeconds))
            print("✓ Resuming after rate limit cooldown")
        }

        exit(0)
    } catch {
        if verbose { print("⚠️  Error checking usage: \(error.localizedDescription)") }
        exit(0)
    }
}

RunLoop.main.run()

File: rate_limit_guard.js
#!/usr/bin/env node
/**
 * Cross-platform rate limit guard for Claude Code.
 * Queries the Anthropic API for rate limit headers using Claude Code's
 * own credentials (OAuth tokens from keychain/credentials file, or ANTHROPIC_API_KEY).
 *
 * No manual session keys or org IDs needed.
 * No external dependencies required (Node 18+).
 */

const { getAuth, queryLimits, formatReset } = require('./query_limits');

const SESSION_THRESHOLD = 90;
const WEEKLY_THRESHOLD = 95;
const verbose = process.argv.includes('--verbose') || process.argv.includes('-v');
const noSleep = process.argv.includes('--no-sleep');

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function formatUsage(utilization) {
  const pct = utilization * 100;
  return pct === Math.floor(pct) ? pct.toString() : pct.toFixed(1);
}

async function main() {
  if (process.env.CLAUDE_NO_LIMIT === '1') {
    if (verbose) console.log('Rate limit guard bypassed (CLAUDE_NO_LIMIT=1)');
    process.exit(0);
  }

  try {
    const auth = getAuth();
    const info = await queryLimits(auth, verbose);

    const sessionClaim = info.claims['5h'];
    const weeklyClaim = info.claims['7d'];

    const sessionUsage = sessionClaim ? sessionClaim.utilization * 100 : 0;
    const weeklyUsage = weeklyClaim ? weeklyClaim.utilization * 100 : 0;

    const sessionStr = formatUsage(sessionClaim ? sessionClaim.utilization : 0);
    const weeklyStr = formatUsage(weeklyClaim ? weeklyClaim.utilization : 0);

    if (verbose) {
      console.log(`Session: ${sessionStr}% (threshold: ${SESSION_THRESHOLD}%) | Weekly: ${weeklyStr}% (threshold: ${WEEKLY_THRESHOLD}%)`);
    }

    let shouldSleep = false;
    let sleepReason = '';
    let resetEpoch = null;

    if (sessionUsage >= SESSION_THRESHOLD) {
      shouldSleep = true;
      sleepReason = `session at ${sessionStr}%`;
      resetEpoch = sessionClaim?.reset;
    } else if (weeklyUsage >= WEEKLY_THRESHOLD) {
      shouldSleep = true;
      sleepReason = `weekly at ${weeklyStr}%`;
      resetEpoch = weeklyClaim?.reset;
    }

    if (shouldSleep && !noSleep) {
      let sleepSeconds = 600; // default 10 minutes

      if (resetEpoch) {
        const secondsUntilReset = Math.floor(resetEpoch - Date.now() / 1000) + 60;
        if (secondsUntilReset > 0) {
          sleepSeconds = secondsUntilReset;
        }
      }

      const minutes = Math.floor(sleepSeconds / 60);
      const hours = Math.floor(minutes / 60);
      const timeStr = hours > 0 ? `${hours}h ${minutes % 60}m` : `${minutes}m`;

      console.log(`Claude ${sleepReason} - sleeping ${timeStr} until reset...`);

      await sleep(sleepSeconds * 1000);
      console.log('Resuming after rate limit cooldown');
    }

    process.exit(0);

  } catch (e) {
    if (verbose) console.log(`Error checking usage: ${e.message}`);
    process.exit(0);
  }
}

main();
